x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopTrainedModelDeploymentAction.java

```diff
@@ -0,0 +1,201 @@
+import org.elasticsearch.xpack.core.ml.action.GetTrainedModelsAction;
+import org.elasticsearch.xpack.core.ml.action.StopTrainedModelDeploymentAction;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelConfig;
+import org.elasticsearch.xpack.core.ml.inference.deployment.TrainedModelDeploymentState;
+import org.elasticsearch.xpack.core.ml.inference.deployment.TrainedModelDeploymentTaskState;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.inference.deployment.TrainedModelDeploymentTask;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+public class TransportStopTrainedModelDeploymentAction extends TransportTasksAction<TrainedModelDeploymentTask,
+    StopTrainedModelDeploymentAction.Request, StopTrainedModelDeploymentAction.Response, StopTrainedModelDeploymentAction.Response> {
+
+    private static final Logger logger = LogManager.getLogger(TransportStopTrainedModelDeploymentAction.class);
+
+    private final Client client;
+    private final ThreadPool threadPool;
+    private final PersistentTasksService persistentTasksService;
+
+    @Inject
+    public TransportStopTrainedModelDeploymentAction(String actionName, ClusterService clusterService, TransportService transportService,
+                                                     ActionFilters actionFilters, Client client, ThreadPool threadPool,
+                                                     PersistentTasksService persistentTasksService) {
+        super(actionName, clusterService, transportService, actionFilters, StopTrainedModelDeploymentAction.Request::new,
+            StopTrainedModelDeploymentAction.Response::new, StopTrainedModelDeploymentAction.Response::new, ThreadPool.Names.SAME);
+        this.client = client;
+        this.threadPool = threadPool;
+        this.persistentTasksService = persistentTasksService;
+    }
+
+    @Override
+    protected void doExecute(Task task, StopTrainedModelDeploymentAction.Request request,
+                             ActionListener<StopTrainedModelDeploymentAction.Response> listener) {
+        ClusterState state = clusterService.state();
+        DiscoveryNodes nodes = state.nodes();
+        if (nodes.isLocalNodeElectedMaster() == false) {
+            redirectToMasterNode(nodes.getMasterNode(), request, listener);
+            return;
+        }
+
+        logger.debug("[{}] Received request to undeploy", request.getId());
+
+        ActionListener<GetTrainedModelsAction.Response> getModelListener = ActionListener.wrap(
+            getModelsResponse -> {
+                List<TrainedModelConfig> models = getModelsResponse.getResources().results();
+                if (models.isEmpty()) {
+                    listener.onResponse(new StopTrainedModelDeploymentAction.Response(true));
+                    return;
+                }
+                if (models.size() > 1) {
```

Comment:
In future we may have more than one model config using the deployment. I might be that we don't do the `GetTrainedModelsAction` here and just look for persistent tasks that match the model ID

Reply:
Agreed. That's how we typically implement stop actions where we handle stopping multiple tasks at once. Just thought this was simpler for now.